Implement support for new epoch columns to StreamDBClient queries

Tiihott opened the pull request:
## Description

The aim of this pull request is to implement support for new epoch time columns to logfile metadata queries done by StreamDBClient. Original logdate and logtime implementations are prone to timezone issues where the result of the query changes depending on the session timezone, using epoch time resolves this issue.

Includes:
- Added support for using epoch_hour column as a source for logtime result value.
- Removed support for using old logdate and logtime implementation.
- New tests for asserting the query changes work as expected.
- Renamed and disabled old logdate implementation tests.
- Fixed weightedOffsetTest which was not using whole hours for logtime values.

Resolves #213
Resolves #214 
Resolves #284

## Checklists

<!-- Fill check boxes before submitting the pull request. -->

### Testing

#### General

- [x] I have checked that my test files and functions have meaningful names.
- [x] I have checked that each test tests only a single behavior.
- [x] I have done happy tests.
- [x] I have tested only my own code.
- [x] I have tested at least all public methods. 

#### Assertions

- [x] I have checked that my tests use assertions and not runtime overhead.
- [x] I have checked that my tests end in assertions.
- [x] I have checked that there is no comparison statements in assertions.
- [x] I have checked that assertions are in tests and not in helper functions.
- [x] I have checked that assertions for iterables are outside of for loops and both sides of the iteration blocks.
- [x] I have checked that assertions are not tested inside consumers.

#### Testing Data

- [x] I have tested algorithms and anything else with the possibility of unbound growth.
- [x] I have checked that all testing data is local and fully replaceable or reproducible or both.
- [x] I have checked that all test files are standalone.
- [x] I have checked that all test-specific fake objects and classes are in the test directory.
- [x] I have checked that my tests do not contain anything related to customers, infrastructure or users.
- [x] I have checked that my tests do not contain non-generic information.
- [x] I have checked that my tests do not do external requests and are not privately or publicly routable.

#### Statements

- [x] I have checked that my tests do not use throws for exceptions.
- [x] I have checked that my tests do not use try-catch statements.
- [x] I have checked that my tests do not use if-else statements.

#### Java

- [x] I have checked that my tests for Java uses JUnit library.
- [x] I have checked that my tests for Java uses JUnit utilities for parameters.

#### Other

- [x] I have only tested public behavior and not private implementation details.
- [x] I have checked that my tests are not (partially) commented out.
- [x] I have checked that hand-crafted variables in assertions are used accordingly.
- [x] I have tested [Object Equality](https://docs.oracle.com/javase/6/docs/api/java/lang/Object.html#equals%28java.lang.Object%29).
- [x] I have checked that I do not have any manual tests or I have a valid reason for them and I have explained it in the PR description.

### Code Quality

- [x] I have checked that my code follows metrics set in Procedure: Class Metrics.
- [x] I have checked that my code follows metrics set in Procedure: Method Metrics.
- [x] I have checked that my code follows metrics set in Procedure: Object Quality.
- [x] I have checked that my code does not have any NULL values.
- [x] I have checked my code does not contain FIXME or TODO comments.

Tiihott committed:
Implemented coalesce with epoch_hour column usage for synthetic logtime field creation. REBASE

Implemented coalesce with epoch_hour converted to date for defining logdate of the logfile. REBASE

Tiihott committed:
Fixed pullToSliceTable to allow using epoch_hour for Date filtering when available. REBASE

Tiihott committed:
Implemented epochHourTest and epochHourNullTest. REBASE

Tiihott committed:
Fixed timezone issues with logdate generation from epoch_hour field using date arithmetics that completely ignores session timezone. Added epochHourTimezoneTest. REBASE

StrongestNumber9 reviewed (changes requested):
The tests requires quite a lot of cleaning and adding the testcontainers as they are very mentally expensive to parse, especially with the duplicated codeblocks

Tiihott commented on src/test/java/com/teragrep/pth_06/planner/StreamDBClientTest.java:
```diff
@@ -0,0 +1,463 @@
+                                )
+                )
+                .execute();
+
```
Moved all the test data insertions that are not directly relevant to the tests to beforeEach in commit 2fbb4de2d4408f3c8dea2308b5ef0d811ca737f0
Created a private method for initializing Config object with test options in commit 1f2b0a68ab1a8472ae91925d99bdd00a31bcc1af

Tiihott commented on src/test/java/com/teragrep/pth_06/planner/StreamDBClientTest.java:
```diff
@@ -0,0 +1,463 @@
+        long logtime = hourRange.get(0).get(8, Long.class);
+        Assertions.assertEquals(1696471200L, logtime);
+        Date logdate = hourRange.get(0).get(5, Date.class);
+        Assertions.assertEquals(Date.valueOf(date), logdate);
```
Changed Date.valueod() to literal string format instead of LocalDate in commit 8f96a57c412eafcf186ea0879624aa294e0803ad

Tiihott commented on src/test/java/com/teragrep/pth_06/planner/StreamDBClientTest.java:
```diff
@@ -0,0 +1,463 @@
+        HostRecord hostRecord = new HostRecord(UShort.valueOf(1), "testHost1");
+        ctx.insertInto(JOURNALDB.HOST).set(hostRecord).execute();
+
+        // Set logdate to 2023-10-04 instead of the correct 2023-10-05 to emulate timezone issues, and test if epoch_hour takes priority or not.
```
Changed the test configuration to use predetermined UTC-4 timezone for all tests instead of whatever the system the test runs on decides to use as session timezone.
epochHourNullTest() now tests if the results are affected by the UTC-4 session timezone as expected when old logdate/logtime implementation is used.
epochHourTimezoneTest() tests if logtime and logdate results are unaffected by the UTC-4 timezone.

Tiihott commented:
Implemented testcontainers to the tests. As testcontainers do not support managing several databases inside the same testcontainer using the methods available on the testcontainer object, the streamdb database had to be initialized using a separate CREATE_STREAMDB_DB.sql that is called during container startup.

StrongestNumber9 commented:
I was able to make it work with podman, opened ticket a ticket for enhancing documentation in https://github.com/teragrep/mvn_01/issues/84

I think this looks acceptable but I would like @kortemik to give another sanity check

Tiihott commented:
Rebased to main.

elliVM reviewed (changes requested):
Some comments

Tiihott committed:
Removed wildcards from imports. REBASE

Tiihott committed:
Added connection closing to cleanup.

Tiihott commented on src/test/java/com/teragrep/pth_06/planner/StreamDBClientTest.java:
Added `connection.close()` to test cleanup in commit 406864538345b1228973b3c8ef9c64a5dd853093

Tiihott committed:
Implemented the usage of two different session timezones to epochHourTimezoneTest().

elliVM reviewed (dismissed):
LGTM!

kortemik commented:
please rebase

Tiihott committed:
Removed support for old logtime source, only epoch_hour supported. REBASE

Tiihott committed:
Added new epoch tests to replace old null epoch tests that were renamed and disabled.

Tiihott committed:
Added second bucket to test data.

Tiihott committed:
Fixed test setup in weightedOffsetTest() and weightedOffsetNullEpochTest().

Tiihott committed:
Added setIncludeBeforeEpochTest() to replace setIncludeBeforeEpochNullEpochTest().

Tiihott committed:
Added deleteRangeFromSliceTableTest() to replace deleteRangeFromSliceTableNullEpochTest().

Tiihott commented:
Rebased and removed the functionality for using old logtime implementation for now.
Renamed and disabled tests that were using the old logtime implementation, and added new epoch logtime tests to replace them.
Updated PR description with the changes.

elliVM commented on src/main/java/com/teragrep/pth_06/planner/NestedTopNQuery.java:
I think that `private final Field<Long> logtime` should be renamed to `epochHour` to imply that it used `epoch_hour`. also `private final Field<Long> logtimeForOrderBy` should be something like `epochHourForOrderBy`.

elliVM commented on src/test/java/com/teragrep/pth_06/planner/StreamDBClientTest.java:
```diff
@@ -180,14 +187,11 @@ private LogfileRecord logfileRecordForEpoch(long epoch, boolean hasNullEpochColu
+        LogfileRecord logfileRecord2 = logfileRecordForEpoch(1696471200L + 24L * 3600L, false);
         ctx.insertInto(JOURNALDB.LOGFILE).set(logfileRecord2).execute();
 
         // Assert StreamDBClient methods work as expected with the test data.
```
options for test cases should be copied from the global opts instead of assigment since this mutates the shared state. 

Instead, do:
`final Map<String, String> opts = new HashMap<>(this.opts)`
and then add options for the test case, this should be done for all test cases.

Tiihott committed:
Renamed logtime usage to epochHour.

Tiihott committed:
Refactored Config options to be copied from the global options in tests.

elliVM reviewed (dismissed):
Looks good to me, question still open about switching to epoch range based query logic. Perhaps can be done in another issue

Tiihott committed:
Fixed all relevant tests to use epoch for logtime instead of the path string.

Tiihott committed:
Fixed comments to reflect the usage of epoch_hour.

Tiihott committed:
Removed logdate from StreamDBClient query results. Refactored all usage of Result10 to Result9 with the Date result value removed.

Tiihott committed:
Replaced logdate and logtime implementation in earliest and latest condition with logfile.epoch_hour implementation. REBASE

Tiihott committed:
Replaced largeNegativeNumberInputTest() with negativeNumberInputExceptionTest().

Tiihott committed:
Added earliest and latest condition tests to StreamDBClientTest.java

Tiihott committed:
Added final keywords to all StreamDBClientTest tests.

elliVM commented on src/main/java/com/teragrep/pth_06/planner/NestedTopNQuery.java:
I think here we still force an index that uses the old logdate, should it be removed?

elliVM commented on src/main/java/com/teragrep/pth_06/planner/NestedTopNQuery.java:
```diff
@@ -117,14 +114,19 @@ public Table<Record> getTableStatement(Condition journaldbConditionArg, Date day
         }
 
         logger.debug("NestedTopNQuery.getTableStatement exit");
+        final Field<Date> logdateFunction = DSL
```
I think you could also do this in native jooq like
```
Field<Date> logdateFunction = DSL
                .dateAdd(DSL.inline(Date.valueOf("1970-01-01")), JOURNALDB.LOGFILE.EPOCH_HOUR, DatePart.SECOND)
                .cast(Date.class);
```

elliVM commented on src/main/java/com/teragrep/pth_06/planner/NestedTopNQuery.java:
```diff
@@ -117,14 +114,19 @@ public Table<Record> getTableStatement(Condition journaldbConditionArg, Date day
         return selectOnConditionStep
-                .where(JOURNALDB.LOGFILE.LOGDATE.eq(day).and(journaldbConditionArg))
-                .orderBy(logtimeForOrderBy, JOURNALDB.LOGFILE.ID.asc())
+                .where(logdateFunction.eq(day).and(journaldbConditionArg))
```
I think that SQL cannot use an index on the `logdateFunction.eq(day)` since it has to calculate every row before it knows the actual value for the row.

We could create an `new EpochDayCondition(epochHour)` object to calculate this out in java.

using something like this we could get the date of a given epoch value:
```
long startOfDayEpoch = startOfDay(epochHour);
long endOfDayEpoch = endOfDay(epochHour);

return JOURNALDB.EPOCH_HOUR.between(startOfDayEpoch, endOfDayEpoch);
```
SQL could now optimize using an index
